### PR TITLE
Access Control: move features to Enterprise

### PR DESCRIPTION
**Pull request by alexanderzobnin**
This PR reorganizes Access Control service according to previously discussed changes. OSS Grafana implements a simple version of Access Control (use built-in roles populated in memory) and doesn't have access to the database.

### CODE REVIEW

File: pkg/services/accesscontrol/evaluator/evaluator.go
Comment (gamab) [inline pkg/services/accesscontrol/evaluator/evaluator.go:23]: This is where the magic happens, either SQL or RAM... Nice!

File: pkg/services/accesscontrol/accesscontrol.go
Comment (gamab) [inline pkg/services/accesscontrol/accesscontrol.go:9]: Silly comment, but I would call the interface AccessController

File: pkg/services/accesscontrol/ossaccesscontrol/builtin_roles.go
Comment (gamab) [inline pkg/services/accesscontrol/ossaccesscontrol/builtin_roles.go:7]: Why not use a map here to make o(1) complex the getBuildInRole function?

File: pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go
Comment (gamab) [inline pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:18]: Why having this struct field since it is a package scope variable already (in builtin_roles.go) ?

Comment (gamab) [reply pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:18]: If its necessary then why don't we add builtInRoles too?

File: pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go
Comment (gamab) [inline pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:35]: Shouldn't we log an error if Cfg has not been injected?

File: pkg/services/accesscontrol/ossaccesscontrol/builtin_roles.go
Comment (alexanderzobnin) [reply pkg/services/accesscontrol/ossaccesscontrol/builtin_roles.go:7]: Yes, makes sense, this code just moved from the previous version, so some optimization may be applied.

File: pkg/services/accesscontrol/ossaccesscontrol/builtin_roles.go
Comment (gamab) [inline pkg/services/accesscontrol/ossaccesscontrol/builtin_roles.go:26]:
```suggestion
var builtInRoles = [string]accesscontrol.RoleDTO{
	"grafana:builtin:users:read:self": {
		Version: 1,
		Permissions: []accesscontrol.Permission{
			{
				Permission: "users:read",
				Scope:      "users:self",
			},
			{
				Permission: "users.tokens:list",
				Scope:      "users:self",
			},
			{
				Permission: "users.teams:read",
				Scope:      "users:self",
			},
		},
	},
}
```

File: pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go
Comment (alexanderzobnin) [reply pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:35]: It's only for tests (Cfg always injected in the real life).

File: pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go
Comment (gamab) [inline pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:58]:
```suggestion
				if builtInRole, ok := builtInRoles[builtInRoleName]; ok {
					for _, p := range builtInRole.Permissions {
						permission := p
						permissions = append(permissions, &permission)
					}
				}
```

File: pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go
Comment (gamab) [inline pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol.go:50]: Maybe we could move the content of the if in another function: `func appendBuildInRolesPermissions(permissions, builtInRoleNames) []*accesscontrol.Permission`. Your call, the function is not that big but has a lot of for/if/for/if/for encapsulated.

File: pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol_test.go
Comment (gamab) [inline pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol_test.go:29]: As discussed on the phone, is it needed for this test ?

**Review by gamab — COMMENTED** [review]
I left a couple of comments, maybe we can have a look at them before merging ?

File: pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol_test.go
Comment (gamab) [inline pkg/services/accesscontrol/ossaccesscontrol/ossaccesscontrol_test.go:73]: We'll need to add failing testcases as well

**Review by gamab — APPROVED** [review]

File: pkg/services/accesscontrol/accesscontrol.go
Comment (vtorosyan) [reply pkg/services/accesscontrol/accesscontrol.go:10]: My two cents input -> `AccessControl` name comes with good enough semantic information and gives a clear message about the single concept it wants to encapsulate, which is "everything you need as a client to work with the access control system". `AccessController` is on the other side is more confusing (and overused, e.g. good old MVC...) term, e.g. it is "I know how to control things and when to control", which is too specific in our case as we want to be open to extension and expansion.

File: pkg/services/accesscontrol/accesscontrol.go
Comment (vtorosyan) [inline pkg/services/accesscontrol/accesscontrol.go:11]: Would it make sense to have generic `IsDisabled() bool` instead? As `registry.CanBeDisabled` is quite specific to the registry and comes with certain assumptions (e.g. you call it after init etc).

File: pkg/services/accesscontrol/accesscontrol.go
Comment (vtorosyan) [inline pkg/services/accesscontrol/accesscontrol.go:16]: Do we really need this abstraction? 🤔 Can't we add `Evaluate(ctx context.Context, user *models.SignedInUser, permission string, scope ...string) (bool, error)` to `AccessControl` . It seems `Evaluator` does not really provide anything beside a second level of indirection 🤔

**Review by vtorosyan — APPROVED** [review]
Few nits, other than that looks good!

File: pkg/services/accesscontrol/accesscontrol.go
Comment (alexanderzobnin) [reply pkg/services/accesscontrol/accesscontrol.go:16]: Yes, I thought about that also.

**Review by sakjur — APPROVED** [review]
LGTM

**Pull request merged**
